libpod/container_exec.go

```diff
@@ -848,17 +848,16 @@ func (c *Container) healthCheckExec(config *ExecConfig, timeout time.Duration, s
 		return -1, err
 	}
 
-	if c.state.ExecSessions == nil {
-		c.state.ExecSessions = make(map[string]*ExecSession)
-	}
-	c.state.ExecSessions[session.ID()] = session
-	defer delete(c.state.ExecSessions, session.ID())
-
 	opts, err := prepareForExec(c, session)
 	if err != nil {
 		return -1, err
 	}
 	defer func() {
+		// cleanupExecBundle MUST be called with the parent container locked.
+		if !unlock {
+			c.lock.Lock()
```

Comment:
you cannot ever lock without calling  syncContainer() because the function will call c.save() with an outdated state otherwise which will cause really bad problems.

Reply:
could we also add the `!c.batched` condition?  Otherwise the block here depends on `unlock` being false only when `!c.batched` and that could change without us noticing the breakage of the dependency.